Refactor VariableTypeManual

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30650 Back out "Back out "Back out "Revert D18542342: Boxed variable dispatch"""
* **#30649 Refactor VariableTypeManual**

Operators in VariableTypeManual are now no longer registered against the VariableTypeId key, but they are registered as compound ops. See https://github.com/pytorch/pytorch/issues/30102 for background.

This also requires the non-variable codegen to ignore them and requires removal of VariableMethodStubs.cpp.

So, because function_wrapper.py now also needs to know which ops are manual, instead of having a hard-coded list in gen_variable_type.cpp for ops with manual implementation, we now have a `manual_kernel_registration` flag in native_functions.yaml that disables the registration of operator kernels for this operator (the schema is still registered). Then, we manually register the right kernels for the operator.

Differential Revision: [D18778191](https://our.internmc.facebook.com/intern/diff/D18778191/)

smessmer committed:
Refactor VariableTypeManual

Operators in VariableTypeManual are now no longer registered against the VariableTypeId key, but they are registered as compound ops. See https://github.com/pytorch/pytorch/issues/30102 for background.

This also requires the non-variable codegen to ignore them and requires removal of VariableMethodStubs.cpp.

So, because function_wrapper.py now also needs to know which ops are manual, instead of having a hard-coded list in gen_variable_type.cpp for ops with manual implementation, we now have a `manual_kernel_registration` flag in native_functions.yaml that disables the registration of operator kernels for this operator (the schema is still registered). Then, we manually register the right kernels for the operator.

Differential Revision: [D18778191](https://our.internmc.facebook.com/intern/diff/D18778191/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor VariableTypeManual"


Operators in VariableTypeManual are now no longer registered against the VariableTypeId key, but they are registered as compound ops. See https://github.com/pytorch/pytorch/issues/30102 for background.

This also requires the non-variable codegen to ignore them and requires removal of VariableMethodStubs.cpp.

So, because function_wrapper.py now also needs to know which ops are manual, instead of having a hard-coded list in gen_variable_type.cpp for ops with manual implementation, we now have a `manual_kernel_registration` flag in native_functions.yaml that disables the registration of operator kernels for this operator (the schema is still registered). Then, we manually register the right kernels for the operator.

Differential Revision: [D18778191](https://our.internmc.facebook.com/intern/diff/D18778191/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -130,6 +130,11 @@ def TypedDict(name, attrs, total=True):  # type: ignore
+DEFAULT_SCHEMA_REGISTRATION = CodeTemplate("""\
+.op(torch::RegisterOperators::options()
+  .schema("${schema_string}")
+  .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA))
```
Not at all related to this PR but FYI I still really hate that you have to specify alias analysis at each operator registration site

smessmer replied:
you soon don't have to anymore, Xintao is currently working on this.

ezyang commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -297,6 +298,56 @@ Tensor & detach_(Tensor & self) {
+static auto registry = torch::RegisterOperators()
+  .op(torch::RegisterOperators::options()
+    .schema("aten::resize_(Tensor(a!) self, int[] size, *, MemoryFormat? memory_format=None) -> Tensor(a!)")
+    .impl_unboxedOnlyKernel<decltype(VariableType::resize_), &VariableType::resize_>(TensorTypeId::VariableTensorId)
```
I think it would be helpful for future readers to have an explanation why some of these are registered to `VariableTensorId`, and why others are registered as catchAll

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -338,6 +338,16 @@ your operator, please do. For a few corner cases, enabling this might not compil
+
+With this flag set, we will not generate code to automatically register the C++ operator
+implementation with the dispatcher. This is a workaround for some special cases we need
+to deal with (see VariableTypeManual.cpp) and should only be used rarely.
```
Don't say "this is a workaround for some special cases", describe the actual special case that we are using it for.

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -1260,18 +1267,23 @@ def add_namedtensor_enabled_macro(code):
         abstract = False
         if isinstance(type_method_dispatch, dict):
             abstract = True
+            assert not option['manual_kernel_registration'] # Having manual_kernel_registration for an abstract method doesn't make sense.
```
Please fix the lint

ezyang commented on aten/src/ATen/native/Copy.cpp:
```diff
@@ -163,6 +164,13 @@ Tensor& copy_(Tensor& self, const Tensor& src, bool non_blocking) {
+    .schema("aten::copy_(Tensor(a!) self, Tensor src, bool non_blocking=False) -> Tensor(a!)")
+    .impl_unboxedOnlyCatchAllKernel<decltype(copy_), &copy_>()
+    .aliasAnalysis(AliasAnalysisKind::FROM_SCHEMA))
+  ;
```
Why does `copy_` need manual registration? It seems to me that the normal mechanism will register a catchAll, as required, and then you can manually register the Variable implementation without having to do anything extra.

ezyang replied:
Ditto with `resize` and the other methods for whom we manually register VariableTypeId implementations

smessmer replied:
The normal mechanism will register a catchAll **and** a variable implementation. We need to skip registering the variable implementation from codegen so we can plug in our own. This is why `copy_` and `resize_` have been part of the `MANUAL_IMPLEMENTATIONS` setup in `gen_variable_type.py`, which is removed by this PR and replaced with `manual_kernel_registration`.

ezyang reviewed (approved):
I'm not a big fan of adding another knob to `native_functions.yaml` but I agree it is expedient. Please address my comments before landing.

smessmer committed:
Update on "Refactor VariableTypeManual"


Operators in VariableTypeManual are now no longer registered against the VariableTypeId key, but they are registered as compound ops. See https://github.com/pytorch/pytorch/issues/30102 for background.

This also requires the non-variable codegen to ignore them and requires removal of VariableMethodStubs.cpp.

So, because function_wrapper.py now also needs to know which ops are manual, instead of having a hard-coded list in gen_variable_type.cpp for ops with manual implementation, we now have a `manual_kernel_registration` flag in native_functions.yaml that disables the registration of operator kernels for this operator (the schema is still registered). Then, we manually register the right kernels for the operator.

Differential Revision: [D18778191](https://our.internmc.facebook.com/intern/diff/D18778191/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor VariableTypeManual"


Operators in VariableTypeManual are now no longer registered against the VariableTypeId key, but they are registered as compound ops. See https://github.com/pytorch/pytorch/issues/30102 for background.

This also requires the non-variable codegen to ignore them and requires removal of VariableMethodStubs.cpp.

So, because function_wrapper.py now also needs to know which ops are manual, instead of having a hard-coded list in gen_variable_type.cpp for ops with manual implementation, we now have a `manual_kernel_registration` flag in native_functions.yaml that disables the registration of operator kernels for this operator (the schema is still registered). Then, we manually register the right kernels for the operator.

Differential Revision: [D18778191](https://our.internmc.facebook.com/intern/diff/D18778191/)

[ghstack-poisoned]

smessmer committed:
Update on "Refactor VariableTypeManual"


Operators in VariableTypeManual are now no longer registered against the VariableTypeId key, but they are registered as compound ops. See https://github.com/pytorch/pytorch/issues/30102 for background.

This also requires the non-variable codegen to ignore them and requires removal of VariableMethodStubs.cpp.

So, because function_wrapper.py now also needs to know which ops are manual, instead of having a hard-coded list in gen_variable_type.cpp for ops with manual implementation, we now have a `manual_kernel_registration` flag in native_functions.yaml that disables the registration of operator kernels for this operator (the schema is still registered). Then, we manually register the right kernels for the operator.

Differential Revision: [D18778191](https://our.internmc.facebook.com/intern/diff/D18778191/)

[ghstack-poisoned]

Baranowski reviewed (commented):
I'm struggling to understand how `manual_kernel_registration` should be used and I would appreciate if you could help me understand (perhaps @albanD can also help). I'm now trying to use `manual_kernel_registration` for all 4 overloads of `set_`, as per https://github.com/pytorch/pytorch/issues/33941#issuecomment-604590213.

In order to do that, I had to remove the `dispatch:` lines in `native_functions.yaml` because:

```
            # If we're here, then our native_functions.yaml entry has dispatch configuration.
            # Having manual kernel registration doesn't make sense.
```

But then all four overloads still appear in `build/aten/src/ATen/NativeFunctions.h` which leads to linking errors. Should I simply only export the CPU implementations and hide the CUDA and Quantized implementations in an anonymous namespace, like it's done for `resize_` in this PR? Choosing only CPU implementation to be exported seems arbitrary to me but my understanding of codegen, dispatch, etc. is still very hazy.

ezyang commented:
@Baranowski can you please post your full PR so we can look at it?

Baranowski commented:
Sure, here is a WiP PR: https://github.com/pytorch/pytorch/pull/35679 I went ahead and placed the CPU implementations in the `at::native::` namespace, while hiding CUDA and Quantized implementations in anonymous namespaces.
